### PR TITLE
Add some unit test samples for commands component

### PR DESCRIPTION
**Pull request by ShenChen-Amazon**
Add some unit test samples for commands component

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -54,13 +56,15 @@ function registerCommands(context: vscode.ExtensionContext): void {
...
     registerWebviews(profileManager, createSkill);
-
-    apiRegisterCommands(context, [
+    ext.askGeneralCommands = [
         new ListSkillsCommand(), new OpenWorkspaceCommand(), new OpenUrlCommand(),
         new InitCommand(profileManager), new GetToolkitInfoCommand(),
         new ViewAllSkillsCommand(), new CreateSkillCommand(createSkill),
         new CloneSkillCommand(), new ChangeProfileCommand(), new AccessTokenCommand(),
-        new DebugAdapterPathCommand(), new CloneSkillFromConsoleCommand()]);
+        new DebugAdapterPathCommand()
```
Comment (wyajunjun) [inline src/extension.ts:64]: Could you check why delete `CloneSkillFromConsoleCommand` here?

File: test/testUtilities.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    public subscriptions: Array<{ dispose(): any }> = [];
+    public workspaceState: vscode.Memento = new FakeMemento();
+    public globalState: vscode.Memento = new FakeMemento();
+    public storagePath: string | undefined;
+    public globalStoragePath = '.';
+    public logPath = '';
+    public extensionMode;
+    public extensionUri;
+    public environmentVariableCollection;
+
```
Comment (wyajunjun) [inline test/testUtilities.ts:22]: the context has another three fields as we found yesterday - storageUri, globalStorageUri, logUri, probably need to be added here.

File: test/runtime/utils/profileHelper.test.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        sandbox = sinon.createSandbox();
+    });
+
+    afterEach(() => {
+        sandbox.restore();
+    });
+
+    describe("Function listExistingProfileNames tests", () => {
+        it("Should return null when askConfig not exist", () => {
+            sandbox.stub(fs, "existsSync").returns(false);
```
Comment (wyajunjun) [inline test/runtime/utils/profileHelper.test.ts:22]: One question about the code style, should the test files keep the same style or not? According to this https://mochajs.org/#arrow-functions, I probably will not use arrow functions.

File: .eslintrc.js [hunk 4]
```diff
@@ -103,6 +103,10 @@ module.exports = {
         "@typescript-eslint/no-unsafe-assignment": "off",
         // TODO: reEnable no-unsafe-member-access
         "@typescript-eslint/no-unsafe-member-access": "off",
+        "@typescript-eslint/restrict-template-expressions": "off",
+        "@typescript-eslint/no-unused-vars": "off",
+        // TODO: reEnable explicit-module-boundary-types
```
Comment (nikhilym) [inline .eslintrc.js:108]: Reason to disable this rule?

File: src/extension.ts (hunk 1 above)
Comment (nikhilym) [reply src/extension.ts:64]: Guess this PR needs a rebase from the master, to pick up the latest changes.

File: src/extensionGlobals.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { ExtensionContext } from "vscode";
+
+import { GenericCommand } from "../src/runtime/lib/API";
+
+/**
+ * Namespace for common variables used globally in the extension.
+ * All variables here must be initialized in the activate() method of extension.ts
+ */
+export namespace ext {
```
Comment (nikhilym) [inline src/extensionGlobals.ts:9]: Any reason why we are only storing references for commands and not all other resources? Like webviews, tree views etc?

File: src/runtime/lib/utils/profileHelper.ts [hunk 6]
```diff
@@ -144,9 +138,13 @@ function selectVendorId(vendors: Array<model.v1.vendorManagement.Vendor>, contex
 }
 
 export function resolveVendorId(profileName: string): string {
-    let vendorId;
+    let vendorId: string;
     if (profileName === AUTH.PLACEHOLDER_ENVIRONMENT_VAR_PROFILE_NAME) {
+        if (process.env.ASK_VENDOR_ID === undefined) {
+            throw new Error('Please make sure env var ASK_VENDOR_ID is defined.');
```
Comment (nikhilym) [inline src/runtime/lib/utils/profileHelper.ts:144]: Is this being handled in all the methods that are calling this function? We should also raise an `AskError` with proper logging if this leads to a failure in toolkit operations.

File: src/utils/commands/openWorkspace.ts [hunk 7]
```diff
@@ -1,19 +1,16 @@
...
 import { Logger } from '../../logger';
 
-
 export class OpenWorkspaceCommand extends AbstractCommand<void> {
     constructor() {
         super('ask.container.openWorkspace');
     }
 
-    async execute(context: CommandContext): Promise<void> {
+    async execute(): Promise<void> {
```
Comment (nikhilym) [inline src/utils/commands/openWorkspace.ts:13]: Why was this removed? With the current eslint, I don't see any errors being shown on this even though it is not used in the method. Also, the abstract class method has this function variable as a required parameter. By deleting it here, how does typescript make sure that the abstract class methods are actually implemented, since the signature is different?

File: test/askContainer/commands/listSkills.test.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+import * as sinon from "sinon";
+import * as model from "ask-smapi-model";
+import { CustomSmapiClientBuilder } from "ask-smapi-sdk";
+
+import { ListSkillsCommand } from "../../../src/askContainer/commands/listSkills";
+import { TelemetryClient } from "../../../src/runtime/lib/telemetry";
+import { SmapiClientFactory, SmapiResource } from "../../../src/runtime";
+import * as profileHelper from "../../../src/runtime/lib/utils/profileHelper";
+
+describe("Command_listSkills tests", () => {
```
Comment (nikhilym) [inline test/askContainer/commands/listSkills.test.ts:12]: nit: Better to add the whole command name. For eg : `ask.container.listSkills`

File: test/askContainer/commands/listSkills.test.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        .client();
+    before(() => {
+        command = new ListSkillsCommand();
+    });
+    after(() => {
+        command.dispose();
+    });
+    beforeEach(() => {
+        sandbox = sinon.createSandbox();
+        sandbox.stub(TelemetryClient.prototype, "sendData");
```
[the remark this answers is not in the document]
Comment (nikhilym) [reply test/askContainer/commands/listSkills.test.ts:26]: Since this telemetry service methods need stubbing across all commands, maybe we can add this as a utility method and reuse across all command tests?

File: test/runtime/utils/profileHelper.test.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            existsSyncStub.withArgs("foo/.ask/cli_config").returns(true);
+            const mkdirSpy = sandbox.stub(fs, "mkdirSync");
+            const writeFileSpy = sandbox.stub(jsonfile, "writeFileSync");
+
+            createConfigFileIfNotExists();
+            assert.ok(mkdirSpy.notCalled);
+            assert.ok(writeFileSpy.notCalled);
+        });
+    });
+});
```
[the remark this answers is not in the document]
Comment (nikhilym) [reply test/runtime/utils/profileHelper.test.ts:86]: Ideally speaking, we should have tests for all public/exported functions in the utility. That being said, we can definitely add the test cases iteratively. I would like to know @ShenChen-Amazon 's plans for these though.

File: test/testUtilities.ts (hunk 2 above)
Comment (nikhilym) [reply test/testUtilities.ts:22]: Seems like the new fields were added in vscode API 1.49.x. Since this file can go out of sync depending on changes in vscode API version, we should do the following : 

- Rebase this file on new API.
- Add a comment regarding vscode api version being used, so that it is explicit.

Also, how are other extension tests handling this? Are they also maintaining and updating these fake contexts manually on new releases?

File: src/extension.ts (hunk 1 above)
Comment (ShenChen-Amazon) [reply src/extension.ts:64]: Yeah, will rebase

File: test/testUtilities.ts (hunk 2 above)
Comment (ShenChen-Amazon) [reply test/testUtilities.ts:22]: Will rebase on new API and add comment for new api is introduced in which version.

For other extensions, they use package-lock.json to lock the version of vscode dependency(which is git ignored in our repo). I think both solution has pros and cons. 

Keeping the package-lock.json make all developers  to use same dependency version, which make the development experience smooth, it could also avoid some breaking changes in latest dependency version. However, by not keeping package-lock.json, we could develop with the latest dependency version.

After comparing, I personally vote for keeping package-lock.json. Mike also keep it in redfox repo.

File: src/extensionGlobals.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    export let context: ExtensionContext;
```
[the remark this answers is not in the document]
Comment (ShenChen-Amazon) [reply src/extensionGlobals.ts:10]: Good catch, will get it assigned in the beginning of actiave function

File: test/askContainer/commands/listSkills.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (ShenChen-Amazon) [reply test/askContainer/commands/listSkills.test.ts:26]: will add stubTelemetryClient util function to testUtilities files, which could help stub 'sendDate' and 'startAction' functions in TelemetryClient.

File: test/runtime/utils/profileHelper.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (ShenChen-Amazon) [reply test/runtime/utils/profileHelper.test.ts:86]: Yeah, since this PR is just a test samples thus I haven't covered all functions in that file. Will add in future PRs.

File: .eslintrc.js (hunk 4 above)
Comment (ShenChen-Amazon) [reply .eslintrc.js:108]: For no-unused-vars, If we turn on this rule, we need to remove all unused vars in function signature. I remembered once you mentioned there are some params in function signature which is not currently used in logic, but will be used in future, thus you leave them in function signature for now. Also, it's related to anther comment you gave below, while implementing an abstract class's method, this lint rule will report error if param declared but not used. 

For restrict-template-expressions, I think we need to bring it back for sure. There are some lint error in current code base related to this rule, the fix would need some context such as: https://github.com/alexa/ask-toolkit-for-vscode/blob/master/src/utils/urlHandlers.ts#L57. I was thinking to disable it for now, thus developer contributor won't be bothered by current lint issues and at the same time we could continually fix all lint rules. Probably add a TODO as well ?

File: src/extensionGlobals.ts (hunk 5 above)
Comment (ShenChen-Amazon) [reply src/extensionGlobals.ts:9]: Yeah, will add them in my next PR for adding tests for trees and webViews.

File: src/utils/commands/openWorkspace.ts (hunk 7 above)
Comment (ShenChen-Amazon) [reply src/utils/commands/openWorkspace.ts:13]: Related to the lint rule I turned off above. Did some research: https://stackoverflow.com/questions/50266006/how-to-avoid-spurious-unused-parameter-warnings-in-typescript. But seems not a good workaround to follow ? We could discuss this one with other folks :)

File: test/askContainer/commands/listSkills.test.ts (hunk 8 above)
Comment (ShenChen-Amazon) [reply test/askContainer/commands/listSkills.test.ts:12]: Good point, will apply to other tests.

File: src/runtime/lib/utils/profileHelper.ts (hunk 6 above)
Comment (ShenChen-Amazon) [reply src/runtime/lib/utils/profileHelper.ts:144]: Ummm, cause I saw this function already have branch to throw error: https://github.com/alexa/ask-toolkit-for-vscode/blob/master/src/runtime/lib/utils/profileHelper.ts#L156, thus I haven't checked carefully at caller side. After checking, seems not all caller side will throw AskError, will try to fix them.

File: test/testUtilities.ts (hunk 2 above)
Comment (nikhilym) [reply test/testUtilities.ts:26]: Let's keep the `package-lock.json` then. It is better to have a smoother, consistent experience across all devs working on the extension, than having different versions in individual machines leading to different errors/experience changes.

File: .eslintrc.js (hunk 4 above)
Comment (nikhilym) [reply .eslintrc.js:108]: A TODO would be a good start.

File: src/askContainer/webViews/createSkillWebview.ts [hunk 12]
```diff
@@ -118,7 +118,12 @@ export class CreateSkillWebview extends AbstractWebView {
         Logger.verbose(`Calling method: ${this.viewId}.createSmapiResource, args: `, skillId, skillName);
         let profile = Utils.getCachedProfile(this.extensionContext);
         profile = profile ?? DEFAULT_PROFILE;
-        const vendorId = Utils.resolveVendorId(profile);
+        let vendorId: string;
+        try {
+            vendorId = Utils.resolveVendorId(profile);
+        } catch (err) {
+            throw loggableAskError(`Failed to retrieve vendorID for profile ${profile}`, err, true);
```
Comment (nikhilym) [inline src/askContainer/webViews/createSkillWebview.ts:125]: nit: Instead of doing this, we can handle the same behavior in the utility function right? Also, I think this is an edge case and is not in general skill dev flow.

**Review by nikhilym — APPROVED** [review]
nit comment but this looks like a good start. 🚢

File: test/testUtilities.ts (hunk 2 above)
Comment (ShenChen-Amazon) [reply test/testUtilities.ts:26]: Got it, will keep package-lock

File: .eslintrc.js (hunk 4 above)
Comment (ShenChen-Amazon) [reply .eslintrc.js:108]: Got it, will add a TODO

**Review by wyajunjun — APPROVED** [review]

**Review by nikhilym — APPROVED** [review]
:ship: shipit:

**Review by wyajunjun — APPROVED** [review]
